code/modules/client/preferences2/preferences2.dm

```diff
@@ -0,0 +1,224 @@
+	pda_color		= sanitize_hexcolor(pda_color, 6, TRUE, initial(pda_color))
+
+	key_bindings 	= sanitize_islist(key_bindings, deepCopyList(GLOB.keybinding_list_by_key))
+	if (!key_bindings)
+		key_bindings = deepCopyList(GLOB.keybinding_list_by_key)
+
+	if(!purchased_gear)
+		purchased_gear = list()
+
+	return TRUE
+
+
+// OH BOY MORE MACRO ABUSE
+#define PREP_WRITEPREF_RAW(value, tag) write_queries += SSdbcore.NewQuery("INSERT INTO [format_table_name("preferences")] (ckey, preference_tag, preference_value) VALUES (:ckey, :ptag, :pvalue) ON DUPLICATE KEY UPDATE preference_value=:pvalue2", list("ckey" = parent.ckey, "ptag" = tag, "pvalue" = value, "pvalue2" = value))
+#define PREP_WRITEPREF_JSONENC(value, tag) PREP_WRITEPREF_RAW(json_encode(value), tag)
+
+
+// Writes all prefs to the DB
+/datum/preferences/proc/save_preferences()
+	if(!SSdbcore.IsConnected())
+		return
+
+	if(IS_GUEST_KEY(parent.ckey))
+		return
+
+	var/list/datum/DBQuery/write_queries = list() // do not rename this you muppet
+
+	//general preferences
+	PREP_WRITEPREF_RAW(default_slot, PREFERENCE_TAG_DEFAULT_SLOT)
+	PREP_WRITEPREF_RAW(chat_toggles, PREFERENCE_TAG_CHAT_TOGGLES)
+	PREP_WRITEPREF_RAW(toggles, PREFERENCE_TAG_TOGGLES)
+	PREP_WRITEPREF_RAW(toggles2, PREFERENCE_TAG_TOGGLES2)
+	PREP_WRITEPREF_RAW(clientfps, PREFERENCE_TAG_CLIENTFPS)
+	PREP_WRITEPREF_RAW(parallax, PREFERENCE_TAG_PARALLAX)
+	PREP_WRITEPREF_RAW(pixel_size, PREFERENCE_TAG_PIXELSIZE)
+	PREP_WRITEPREF_RAW(tip_delay, PREFERENCE_TAG_TIP_DELAY)
+	PREP_WRITEPREF_RAW(pda_style, PREFERENCE_TAG_PDA_STYLE)
+	PREP_WRITEPREF_RAW(pda_color, PREFERENCE_TAG_PDA_COLOUR)
+
+	PREP_WRITEPREF_RAW(asaycolor, PREFERENCE_TAG_ASAY_COLOUR)
+	PREP_WRITEPREF_RAW(ooccolor, PREFERENCE_TAG_OOC_COLOUR)
+	PREP_WRITEPREF_RAW(lastchangelog, PREFERENCE_TAG_LAST_CL)
+	PREP_WRITEPREF_RAW(UI_style, PREFERENCE_TAG_UI_STYLE)
+	PREP_WRITEPREF_RAW(outline_color, PREFERENCE_TAG_OUTLINE_COLOUR)
+	PREP_WRITEPREF_RAW(see_balloon_alerts, PREFERENCE_TAG_BALLOON_ALERTS)
+	PREP_WRITEPREF_RAW(scaling_method, PREFERENCE_TAG_SCALING_METHOD)
+	PREP_WRITEPREF_RAW(ghost_form, PREFERENCE_TAG_GHOST_FORM)
+	PREP_WRITEPREF_RAW(ghost_orbit, PREFERENCE_TAG_GHOST_ORBIT)
+	PREP_WRITEPREF_RAW(ghost_accs, PREFERENCE_TAG_GHOST_ACCS)
+	PREP_WRITEPREF_RAW(ghost_others, PREFERENCE_TAG_GHOST_OTHERS)
+
+	PREP_WRITEPREF_JSONENC(ignoring, PREFERENCE_TAG_IGNORING)
+	PREP_WRITEPREF_JSONENC(key_bindings, PREFERENCE_TAG_KEYBINDS)
+	PREP_WRITEPREF_JSONENC(purchased_gear, PREFERENCE_TAG_PURCHASED_GEAR)
+	PREP_WRITEPREF_JSONENC(be_special, PREFERENCE_TAG_BE_SPECIAL)
+
+	// QuerySelect can execute many queries at once. That name is dumb but w/e
+	SSdbcore.QuerySelect(write_queries, TRUE, TRUE)
+
+
+// Get ready for a disgusting SQL query
+/datum/preferences/proc/load_characters()
+	// Do NOT remove stuff from the start of this query. Only append to the end.
+	// If you delete an entry, god help you as you have to update all the indexes
```

Comment:
Can you explain the reasoning behind this? The only index I see is slotckey, which shouldn't care about the order of the other entries.

Reply:
Its easy to forget about updating all these old ones, then you get offsets on what data should be where
![image](https://user-images.githubusercontent.com/25063394/176913333-ae0c91ec-a423-4b1c-995a-fef55608b1ba.png)

Reply:
Ah, maybe tweak the comment to "the indexes in handle_query"? That way if any poor soul does need to do it (because, say, they're removing a property) they know where to go, and also it won't get confused with DB indexes.